metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -269,9 +270,237 @@ void TaxonomyClsAnno::rmq_preprocessing(const std::vector<TaxId> &tree_lineariza
+        if (!nodes_already_propagated->count(act_node)) {
+            if (num_kmers_per_node.count(act_node)) {
+                score_from_unprocessed_parents += num_kmers_per_node.at(act_node);
+            }
+            unprocessed_parents.push_back(act_node);
+        } else {
+            if (num_kmers_per_node.count(act_node)) {
+                score_from_processed_parents += num_kmers_per_node.at(act_node);
+            }
+            processed_parents.push_back(act_node);
+        }
+    }
+    // The score of all the nodes in 'processed_parents' will be updated with 'score_from_unprocessed_parents' only.
+    // The nodes in 'unprocessed_parents' will be updated with the sum 'score_from_processed_parents + score_from_unprocessed_parents'.
+    for (uint64_t i = 0; i < unprocessed_parents.size(); ++i) {
+        TaxId &act_node = unprocessed_parents[i];
+        (*node_scores)[act_node] =
+                score_from_processed_parents + score_from_unprocessed_parents;
+        nodes_already_propagated->insert(act_node);
+
+        uint64_t act_dist_to_root =
+                processed_parents.size() + unprocessed_parents.size() - i;
+
+        // Test if the current node's score would be a better LCA result.
+        if ((*node_scores)[act_node] >= desired_number_kmers
+            && (act_dist_to_root > *best_lca_dist_to_root
+                || (act_dist_to_root == *best_lca_dist_to_root
+                    && (*node_scores)[act_node] > (*node_scores)[*best_lca])
+                )
+            ) {
+            *best_lca = act_node;
+            *best_lca_dist_to_root = act_dist_to_root;
+        }
+    }
+    for (uint64_t i = 0; i < processed_parents.size(); ++i) {
+        TaxId &act_node = processed_parents[i];
+        (*node_scores)[act_node] += score_from_unprocessed_parents;
+
+        uint64_t act_dist_to_root = processed_parents.size() - i;
+        if ((*node_scores)[act_node] >= desired_number_kmers
+            && (act_dist_to_root > *best_lca_dist_to_root
+                || (act_dist_to_root == *best_lca_dist_to_root
+                    && (*node_scores)[act_node] > (*node_scores)[*best_lca])
+                )
+            ) {
+            *best_lca = act_node;
+            *best_lca_dist_to_root = act_dist_to_root;
+        }
+    }
+}
+
+TaxId TaxonomyClsAnno::find_lca(const std::vector<TaxId> &taxids) const {
+    if (taxids.empty()) {
+        logger->error("Internal error: Can't find LCA for an empty set of normalized taxids.");
+        exit(1);
+    }
+    uint64_t left_idx = node_to_linearization_idx.at(taxids[0]);
+    uint64_t right_idx = node_to_linearization_idx.at(taxids[0]);
+
+    for (const TaxId &taxid : taxids) {
+        if (node_to_linearization_idx.at(taxid) < left_idx) {
+            left_idx = node_to_linearization_idx.at(taxid);
+        }
+        if (node_to_linearization_idx.at(taxid) > right_idx) {
+            right_idx = node_to_linearization_idx.at(taxid);
```

Comment:
so many `.at` ...
get a reference once and re-use it. We've already come across this thing with `.find` vs `.at` ...